### PR TITLE
Multi-signature registration - Create member accounts - Closes #3313

### LINKED ISSUE #3313: Multi-signature registration - Create member accounts
@pablitovicente commented on [Fri Apr 12 2019](https://github.com/LiskHQ/lisk/issues/3306)

### Expected behavior

When I register a multi-signature account I expect the group member accounts to be created

### Actual behavior

When I register a multi-signature account the registration is correct but member accounts are not being created

### Steps to reproduce

Register a multi-signature account and check the mem_account tables for the accounts or call `GET /api/accounts/<address>/multisignature_groups`

### Which version(s) does this affect? (Environment, OS, etc...)

1.7/2.0

### PR DESCRIPTION
**Pull request by pablitovicente**
### What was the problem?

Group member accounts were not being created when registering a multi-signature transaction

### How did I fix it?

By adding methods to fetch/create the group member accounts

### How to test it?

- Register a multi-signature with two members and the two members should have accounts created
- Credit two accounts, register multi-signature for the first one and use the second one and another one as members. Registration should succeed and second account should have same balance

### Review checklist

* The PR resolves #3313
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

File: elements/lisk-transactions/src/4_multisignature_transaction.ts [hunk 1]
```diff
@@ -101,10 +117,15 @@ export class MultisignatureTransaction extends BaseTransaction {
 	}
 
 	public async prepare(store: StateStorePrepare): Promise<void> {
+		const membersAddresses = extractPublicKeysFromAsset(
+			this.asset.multisignature.keysgroup,
+		).map(publicKey => ({ address: getAddressFromPublicKey(publicKey) }));
```
Comment (SargeKhan) [inline elements/lisk-transactions/src/4_multisignature_transaction.ts]: [Address collision](https://github.com/LiskHQ/lisk-sdk/issues/266) is an existing problem in Lisk protocol which may lead to fetching collision accounts if we convert public keys to addresses when querying from the database. Therefore, I would prefer if we use public keys to fetch accounts instead of addresses.

File: elements/lisk-transactions/src/4_multisignature_transaction.ts [hunk 2]
```diff
@@ -65,6 +66,21 @@ export const multisignatureAssetFormatSchema = {
...
+) => {
+	membersPublicKeys.map(memberPublicKey => {
+		const address = getAddressFromPublicKey(memberPublicKey);
+		const memberAccount = { ...store.account.getOrDefault(address) };
+		memberAccount.publicKey = memberAccount.publicKey || memberPublicKey;
+		store.account.set(memberAccount.address, memberAccount);
+	});
+};
+
+const extractPublicKeysFromAsset = (assetPublicKeys: ReadonlyArray<string>) =>
```
Comment (SargeKhan) [inline elements/lisk-transactions/src/4_multisignature_transaction.ts]: Maybe rename this function to extractPublicKeysFromVotes?

Comment (pablitovicente) [reply elements/lisk-transactions/src/4_multisignature_transaction.ts]: this one is for Multi-signatures not votes

File: elements/lisk-transactions/src/4_multisignature_transaction.ts (hunk 1 above)
Comment (SargeKhan) [reply elements/lisk-transactions/src/4_multisignature_transaction.ts]: After discussion face-to-face, we concluded that it's necessary to convert public key. Converting public keys to addresses ensures that uninitialized accounts are also fetched and not overwritten with default account values. 
Nice catch @pablitovicente!

File: elements/lisk-transactions/src/4_multisignature_transaction.ts [hunk 3]
```diff
@@ -65,6 +66,21 @@ export const multisignatureAssetFormatSchema = {
 	},
 };
 
+const setMemberAccounts = (
+	store: StateStore,
+	membersPublicKeys: ReadonlyArray<string>,
+) => {
+	membersPublicKeys.map(memberPublicKey => {
```
Comment (shuse2) [inline elements/lisk-transactions/src/4_multisignature_transaction.ts]: this should be `forEach` since it's not mapping anything?

File: elements/lisk-transactions/src/4_multisignature_transaction.ts [hunk 4]
```diff
@@ -65,6 +66,21 @@ export const multisignatureAssetFormatSchema = {
...
+		const address = getAddressFromPublicKey(memberPublicKey);
+		const memberAccount = { ...store.account.getOrDefault(address) };
```
Comment (shuse2) [inline elements/lisk-transactions/src/4_multisignature_transaction.ts]: assuming `getOrDefault` gives back the copy maybe like
```
const memberAccount = store.account.getOrDefault(address);
const memberAccountWithPublicKey = {
  ...memberAccount,
  publicKey: memberAccount.publicKey || memberPublicKey.
};
```

File: framework/test/mocha/functional/http/get/accounts/multisignatures.js [hunk 5]
```diff
@@ -72,8 +72,7 @@ describe('GET /api/accounts', () => {
 		);
 
 		describe('address', () => {
-			// eslint-disable-next-line
-			it.skip('[feature/improve_transactions_processing_efficiency] using known address should respond with its multisignature_group', async () => {
+			it('using known address should respond with its multisignature_group', async () => {
```
Comment (shuse2) [inline framework/test/mocha/functional/http/get/accounts/multisignatures.js]: maybe rephrase it to
`it should respond with it's multisignature group when using known address`?
same for below

File: framework/test/mocha/functional/http/post/4.multisignature.accounts.js [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+			return expect(res.statusCode).to.be.equal(200);
+		});
+		// Wait for confirmation of credits
+		const transactionsToWaitFor = transactions.map(
+			transaction => transaction.id
+		);
+		await waitFor.confirmations(transactionsToWaitFor);
+	});
+
+	it('When members do not exist in blockchain should be created', async () => {
```
Comment (shuse2) [inline framework/test/mocha/functional/http/post/4.multisignature.accounts.js]: ❤️

File: framework/test/mocha/functional/http/get/accounts/multisignatures.js (hunk 5 above)
Comment (pablitovicente) [reply framework/test/mocha/functional/http/get/accounts/multisignatures.js]: Done
